### PR TITLE
Assistant Cap Bugfixes

### LINKED ISSUE #32749: Warden and Detective do not open Assistant Slots
(WEB REPORT BY: primed REMOTE: 206.221.180.138:7777)
# Revision
cc75c1d055c54a6ff2b29c9b422c45040b7dcb69

# Description
Security roles Detective and Warden do not open new assistant slots while filled. 

# What you Expected
to play assistant
# What Actually Happened
there are 2 security members and yet only 2 assistants allowed max

### LINKED ISSUE #33266: Assistant sent back to lobby due to filled roles could immediately latejoin as one
(WEB REPORT BY: gurfan REMOTE: 206.221.180.138:7777)
# Revision
eca36063df0c9af04779fc53c6e4861be01d4880

# Description
Someone readied up as an assistant was sent back to the lobby due to the roles being filled. Despite this, they could immediately latejoin as an assistant.

# What you Expected
Not to get sent back to the lobby when there was still an open slot.

### LINKED ISSUE #33422: Assistant Cap Roundstart Calculations Bugged
I believe Assistant Cap Calculations are bugged in some way, leading to it returning more people to lobby then it should. This issue involves security being in the game with multiple assistants being returned to lobby, but then being able to latejoin immediately.
I had a general feeling that this was in effect, but I have a greater idea of what's occuring due to a round a few days back:

There were 7 assistants readied, and 0 security set to high. I had assistant set to high, and Warden to Medium. When I failed to roll Assistant, it placed me into the game as Warden with another security officer but it only allowed 2 assistants into the game. Immediately after people could latejoin.
As another example today there was one roundstart detective, but only 2 assistants were let in. People could also latejoin immediately after this.

My guesses are that it determines the cap and lobby returnees too early, but I'm not entirely certain. Regardless, it's annoying.

### PR DESCRIPTION
**Pull request by jwhitak**
# **tl;dr detectives count now and you don't have to latejoin assistant at roundstart anymore**

image

## What this does

To understand what this PR does, we must first learn how the Assistant Cap works. It's divided into three checks.

**AC1**: The initial check. When the job assignment assigner comes to you, and your preference is assistant, this is the first assistant cap check that is triggered.
image

**AC2**: This is the second check. If you are unassigned after all of the job assignments have been issued and you have Join as Assistant if Unassigned enabled, this check is triggered.
image

**AC3**: This is the third check. If you had Assistant as a higher priority job then the one you were assigned, then this check triggers, giving you a chance to be reassigned to Assistant. This would likely happen if you had assistant set to High, but the assistant cap was already reached at THAT TIME. After you got denied, other players were assigned to security (likely from their Medium or Low preferences), increasing the cap. This is a 'second chance' to get your higher preference.
image
(e.g. If there were already 2 assistants and 0 secoffs when your HIGH roll came, you would be rejected normally because of **AC1**. After your rejection, the assignment process would continue and in this example another player would be assigned to security. Then your MEDIUM roll would come and you would get Captain. After all the jobs were assigned, **AC3** would see that you had a chance to be an Assistant but missed it. It would give you a second chance, checking the now-after-everyone-is-assigned cap to see if it made a mistake. In our example, now that there's another security officer, the cap would have increased so you would be reassigned from Captain to Assistant!)

### Now that we know about these three checks...

**AC1** did not check for detectives.

**AC2** works fine. *Edit: But it would not update the number of counted assistants as it added them, so you could, in theory, have like 10 assistants added to a single open slot in this step.

**AC3** did absolutely nothing - it was broken and always failed. *Edit: And the new version had the same problem as AC2.

So with that in mind, this PR:

Fixes #32749. This bugfixes detectives not counting as security officers for the purposes of the assistant cap in **AC1**. Wardens already counted, but likely this player ran into the bug below. #32993 got the **AC2** check fixed, but they missed **AC1**.

Fixes #33422. Fixes #33266. This PR also fixes **AC3**. The function was literally broken. It checked for security officers, discarded that information, and then did a check that was basically if(0). Now it properly works, including factoring in if you've already been assigned to a security role. If you are, it calculates if you should be reassigned based on the cap that would be in effect if you were reassigned.

This PR also adds the Security Officer Count as a standardized proc.

## Why it's good
Now more assistants can spawn without having to latejoin if there's enough security. More assistants means higher chances of a roundstart welderbombing in tool storage. This means that the administration potentially gets to qdel and then ban people more often. Admins qdeling people has a funny sound effect, and that sound effect makes me laugh. So this PR makes more of that sound effect. Therefore, it's good.

## Changelog

:cl:
 * bugfix: The Assistant Cap correctly checks for Detectives.
 * bugfix: The Assistant Cap now gives a second chance for assistants who missed the cap the first time but had more security players get selected after them.

[bugfix] [roleissue]

### CODE REVIEW

**Comment by PrimeDSS13** [comment]
my knight in shining armor...

**Comment by jwhitak** [comment]
This will probably conflict with #33424. I can rebase and resolve the conflicts if it's merged before this one.

**Comment by adacovsk** [comment]
> This will probably conflict with #33424. I can rebase and resolve the conflicts if it's merged before this one.

It shouldn't we don't modify the same lines of code

**Comment by WindowsErrors** [comment]
with each passing day our top mathematicians understand more and more what makes the assistant cap work
